src/StreamJsonRpc/HeaderDelimitedMessageHandler.cs

```diff
@@ -241,12 +244,13 @@ private enum HeaderParseState
             // Transmit the headers.
             // Ignore the cancellation token so we don't write the header without the content.
             sendingBufferStream.Position = 0;
-            await sendingBufferStream.CopyToAsync(this.SendingStream, MaxHeaderElementSize).ConfigureAwait(false);
+            Task headersTask = sendingBufferStream.CopyToAsync(this.SendingStream, MaxHeaderElementSize);
```

Comment:
is this okay to write to the stream concurrently? shouldn't headertask and contenttask needed to be continueWith? what guarantee ordering of headerTask and contentTask. contentTask can run before headerTask.

Reply:
so a caller of this method acquires a lock to guard the stream from being messed up from being called concurrently? 

so, this method guarantee that it only return once message is written to the stream, but it doesn't guarantee that the other side actually got the message? if so, the guarantee that the message actually written to the stream has no significant value to the writer, I think. if so, how about changing design that. this method just write message to pooled stream (such as memory stream you have above) and enqueue that message (stream) to a queue (FIFO queue)
and return. and one dedicated thread or task, empty that queue. that way you basically serialize all message and 1 thread/task completely own the stream (so no messing up)

Reply:
Well, I think you have highlighted a problem: if the header is larger than `MaxHeaderElementSize` then `CopyToAsync` may not be done calling `this.SendingStream.WriteAsync` for all blocks before our `WriteAsync(contentBytes)` call. So we do have a problem here.

Reply:
We already have a FIFO queue (by virtue of the `AsyncSemaphore`). The Task returned by this method indicates when the sending stream has been completely written to.

Reply:
I thought you were trying to improve throughput. using async (either async lock or await or task) can cause big gap between continuation I believe. especially when there are a lot of pending tasks.

using explicit queue + synchronous lock (or concurrent queue) + explicit loop will give you fastest throughput I believe. at the cost of 100% CPU of 1 core if there are a lot of pending messages.

Reply:
I'm trying to improve throughput by removing the requirement to flush while holding the semaphore. We have very significant gains by doing just that.
A dedicated thread that sleeps a lot still waits for the kernel to wake it up. Being async doesn't require the kernel to get involved and a single threadpool thread can go through tasks faster than the kernel can suspend and resume various threads. At least in some contexts async can be faster. But doubling throughput satisfies me for this PR anyway. :) 
Besides, it turns out there are a *lot* of these ServiceHub streams in VS nowadays, and a dedicated thread for each one of them would get a bit much, IMO.

Reply:
I am glad that You already see performance improvements. I am sorry for not being able to provide my sample/test code at the moment, but the one I have integrates with non-public libs.

>A dedicated thread that sleeps a lot still waits for the kernel to wake it up. Being async doesn't require the kernel to get involved and a single threadpool thread can go through tasks faster than the kernel can suspend and resume various threads.

I agree, additional thread could be also a problem when many connections are to be handled using separate (session related/isolated) instances...

One 'idea' would be to mimic behavior of single thread handling the queue using tasks and making sure there is always at least 1 task planned when the action is necessary.

I will definitely evaluate this improved version (currently I use 1.1.102) and I will get back with findings, but no guarantees when this will happen...